CandidateExercise/ConfigurationExercise/Config.cs

```diff
@@ -88,42 +113,166 @@ public void Shutdown()
 
         public ConfigValue GetValue(string key)
         {
-            throw new NotImplementedException();
+            _log.Info("In GetValue");
+            try
+            {
+                var persistenceStorage = _persistSource;
+                KVPList kvList = GetValues(new List<String>() { key });
+                if (kvList == null)
+                    throw new PersistException("cannon get the value " + key, null);
+                return kvList[0].Value;
+            }
+            catch (PersistException ex)
+            {
+                throw catchException("Persist Exception in  GetValue" , ex);
+            }
+            catch (Exception ex)
+            {
+                throw catchException("Exception in  GetValue",ex);
+            }
         }
 
         public void SaveValue(KVP kvp)
         {
-            throw new NotImplementedException();
+            _log.Info("In SaveValue");
+            try
+            {
+                var list = new List<KVP>() { kvp };
+                SaveValues(list);
+            }
+            catch (PersistException ex)
+            {
+                throw catchException("Persist Exception in  SaveValue",ex);
+            }
+            catch (Exception ex)
+            {
+                throw catchException("Exception in  SaveValue",ex);
+            }
         }
 
         public void SaveValues(KVPList kvps)
         {
-            throw new NotImplementedException();
+            _log.Info("In SaveValues");
+            try
+            {
+                _persistSource.PersistValues(kvps);
+            }
+            catch (PersistException ex)
+            {
+                throw catchException("Persist Exception in  SaveValues",ex);
+            }
+            catch (Exception ex)
+            {
+                throw catchException("Exception in  SaveValues", ex);  
+            }
+
         }
 
         public KVPList GetValues(KList keys)
         {
-            throw new NotImplementedException();
+            _log.Info("In GetValues");
+            KVPList kvpList = null;
+            var storage = _persistSource;
```

Comment:
Why did you chose to assign and use this as a local variable? (curiosity only, not a big concern).

Reply:
assignment of instance/global variables to local variables inside the function scope helps me to read the variables and also limits the tampering that can be done with the main data structure

Reply:
Makes sense to me, thanks.